libs/product/state/src/reducers/composite-product-entities/composite-product-entities.reducer.ts

```diff
@@ -49,9 +49,11 @@ export function daffCompositeProductEntitiesReducer<T extends DaffProduct, V ext
       );
     case DaffProductActionTypes.ProductLoadSuccessAction:
     case DaffProductPageActionTypes.ProductPageLoadSuccessAction:
-      if(action.payload.type === DaffProductTypeEnum.Composite) {
+      const product = action.payload.products[action.payload.id];
```

Comment:
Since related and upsell products might be composite, don't you need to do something like: 
```
return adapter.upsertMany(
    action.payload.products.filter(product => product.type === DaffProductTypeEnum.Composite)
);
```
That would be if it was an array anyway. I guess that begs the question: why do we want this to be a dictionary instead of an array?

Reply:
Oh, is it a dictionary because that assumes we'll need to access the main product more often than we'll need to iterate over all products? I'm not sure that's true though

Reply:
> why do we want this to be a dictionary instead of an array

Easy uniqueness assurances

Reply:
So the regular entity state doesn't have a check for simple type. Does that mean that products are duplicated in different entity states? I do think your suggestion is necessary to implement but the multiple entity states are confusing.

Reply:
nvm ignore that ^ I see that the states are just storing the extra fields that aren;t covered by the main entity state

Reply:
changed

Reply:
> > why do we want this to be a dictionary instead of an array
> 
> Easy uniqueness assurances

I think we're given the same uniqueness assurances by entity state. I see `DaffProductResponse` as just a vehicle to get products from the driver call into state. I think it probably makes more sense to use an array in the `DaffProductResponse` instead of a Dictionary, because you're just transforming from an array to a dictionary and then back into an array when the products are loaded into entity state. Then the `upsertMany` converts it back into a dictionary. I'm not even sure there will ever be a time when we'll need to get only the main product (which is when the dictionary will be useful). 

Additionally, if you're worries about uniqueness between the driver call and state, I don't think there will be any difference in practice between using an array vs dictionary. If it is a dictionary and there are two products with the same Id returned from the backend, then the first instance of the id will be overwritten in the driver layer. If it is an array, the overwriting will happen in the entity state layer instead of the driver.

Reply:
That's well reasoned. I'm convinced but what does @damienwebdev think? And do we even try to ensure uniqueness in the products array coming from the driver?

Reply:
changed

Reply:
I agree.